crates/orderbook/src/signature_validator.rs

```diff
@@ -0,0 +1,49 @@
+use anyhow::{Context, Result};
+use ethcontract::Bytes;
+use primitive_types::H160;
+use shared::Web3;
+
+#[cfg_attr(test, mockall::automock)]
+#[async_trait::async_trait]
+/// <https://eips.ethereum.org/EIPS/eip-1271>
+pub trait SignatureValidator: Send + Sync {
+    async fn is_valid_signature(
+        &self,
+        contract_address: H160,
+        hash: [u8; 32],
+        signature: &[u8],
+    ) -> Result<bool>;
+}
+
+pub struct Web3SignatureValidator {
+    web3: Web3,
+}
+
+impl Web3SignatureValidator {
+    /// The Magical value as defined by EIP-1271
+    pub const MAGICAL_VALUE: [u8; 4] = 0x1626ba7e_u32.to_le_bytes();
```

Comment:
Maybe more naturally written as `[0x16, 0x26, ...]` but no strong opinion.

Reply:
I don't have a strong preference for this but I've kept the value the same way as it is written in EIP-1271.

Reply:
I would slightly prefer: `hex!(1626ba7e)` (from the `hex-literal` crate which we already have as a dependency) but also 100% fine as it and no strong opinion either.

Reply:
> I would slightly prefer: `hex!("1626ba7e")` (from the `hex-literal` crate which we already have as a dependency) but also 100% fine as it and no strong opinion either.

hm.... Yeah, I don't mind using `hex!`, one thing to note however would be to reverse the bytes since they are expected in little-endian by EVMs which might get a bit confusing, rather than having `to_le_bytes` explicitly.

Let me know what you prefer so we can continue with the PR/merge

Reply:
I would have assumed that because it is `bytes` type in the eip that the byte order doesn't matter if you don't go through an integer. So with either hex-literal or manual array it should be the same order as in the eip.

Reply:
> one thing to note however would be to reverse the bytes since they are expected in little-endian by EVMs which might get a bit confusing

I thought hex-byte literals were laid out in memory as-is (without any little/big endian consideration).

So
```solitidity
contract C {
    function f() external pure returns (bytes memory b) {
        bytes4 v = 0x01020304;
        b = new bytes(32);
        assembly {
            mstore(add(32, b), v)
        }
    }
}
```

This will return:
```
0x0102030400000000000000000000000000000000000000000000000000000000
```

Which should get decoded to a `[1, 2, 3, 4] as [u8; 4] == hex!("01020304")`, so no need to reverse bytes.

Reply:
To me `hex!` makes the most sense as we can write the literal exactly like in the ERC (its an ERC now and not EIP 🎉) and will produce a literal value of `[u8; 4]` which is what we want right?